### ISSUE TITLE
Support Solidity 0.8.4 Custom Errors

### ISSUE THREAD

**Issue by dmihal** [nomiclabs/hardhat#1428]
Solidity 0.8.4 has introduced "Custom Errors", which are error messages encoded with a signature & data, similar to function calls.

https://blog.soliditylang.org/2021/04/21/custom-errors/

However, Hardhat currently tries to decode all revert data as a string. If custom errors are used, it will throw the following error:

```
Error: Number can only safely store up to 53 bits
    at assert (node_modules/bn.js/lib/bn.js:6:21)
    at BN.toNumber (node_modules/bn.js/lib/bn.js:519:7)
    at decodeSingle (node_modules/ethereumjs-abi/lib/index.js:243:52)
    at decodeSingle (node_modules/ethereumjs-abi/lib/index.js:223:17)
    at Function.ABI.rawDecode (node_modules/ethereumjs-abi/lib/index.js:390:19)
    at Object.decodeRevertReason (node_modules/hardhat/src/internal/hardhat-network/stack-traces/revert-reasons.ts:10:19)
    at HardhatNode._manageErrors (node_modules/hardhat/src/internal/hardhat-network/provider/node.ts:1206:8)
    at HardhatNode._gatherTraces (node_modules/hardhat/src/internal/hardhat-network/provider/node.ts:937:12)
```

Hardhat should parse Error definitions from the contract's ABI and parse "Custom Errors" correctly.

**Comment by vince0656** [comment]
was hoping to use this feature to save GAS over revert strings!

**Comment by frangio** [comment]
Related: https://github.com/nomiclabs/hardhat/issues/1209.

**Comment by fvictorio** [comment]
We just released Hardhat 2.4.0, with support for custom errors. Check the [release notes](https://github.com/nomiclabs/hardhat/releases/tag/hardhat-core-v2.4.0) to learn more.
